feat: Adapt to `visualization-tool-interface`

AlejandroFernandezLuces opened the pull request:
## Overview

This PR adapts PyPrimeMesh to use https://github.com/ansys/ansys-tools-visualization-interface

This should facilitate maintenance of the repo, as well as provide some decoupling from PyVista, and some extra features in the plotter interface.

RobPasMue commented on doc/source/conf.py:
```diff
@@ -8,6 +8,8 @@
 
 from ansys.meshing.prime import __version__
 
+os.environ["PYANSYS_VISUALIZER_DOC_MODE"] = "true"
```
```suggestion
import ansys.tools.visualization_interface as viz_interface

viz_interface.DOCUMENTATION_BUILD = True
```

RobPasMue commented on doc/source/user_guide/graphics.rst:
```diff
@@ -28,7 +29,11 @@ you to limit the display to particular regions of the model:
 
     >>> # display the first part only
-    >>> display(scope=prime.ScopeDefinition(model, part_expression=model.parts[0].name))
+    >>> display = PrimePlotter()
+    >>> display.plot(
+    ...     model, scope=prime.ScopeDefinition(model, part_expression=model.parts[0].name)
+    ... )
+    >>> display.show()
```
Can't we come up with a wrapper ``display(**kwargs)`` function that wraps these three calls around? That way you don't break that much the old API...

RobPasMue replied:
This applies for the entire codebase.

AlejandroFernandezLuces replied:
The current API calls directy the object to show the plotter, rather than calling a method:

```python
display = Graphics(model)
display()
```

This is something I've been wanting to change for a while, I think this is not an intuitive way to use the plotter. 
Another issue with this is they could make changes to the Model and keep plotting different things in the plotter:

```python
for part_name in coarse_wrap:
    # Each open part before wrap
    display(scope=prime.ScopeDefinition(model, part_expression=part_name))
    closed_part = mesh_util.wrap(
        input_parts=part_name,
        max_size=coarse_wrap[part_name],
        remesh_postwrap=False,
        enable_feature_octree_refinement=False,
    )
    # Closed part with no hole
    display(scope=prime.ScopeDefinition(model, part_expression=closed_part.name))

```

So model and mesh information is being kept in the plotter itself. I've implemented the `Mesh` class to avoid this, and have a clear separation between meshes and plotting. Issue is that the API needs to change, as information is managed differently. I think this is more PyVista-like:

```python

for part_name in coarse_wrap:
    # Each open part before wrap
    scope = prime.ScopeDefinition(model, part_expression=part_name)
    pl = PrimePlotter()
    pl.plot(model, scope)
    pl.show()
    closed_part = mesh_util.wrap(
        input_parts=part_name,
        max_size=coarse_wrap[part_name],
        remesh_postwrap=False,
        enable_feature_octree_refinement=False,
    )
    # Closed part with no hole
    scope = prime.ScopeDefinition(model, part_expression=closed_part.name)

    pl = PrimePlotter()
    pl.show(model, scope)
```

Note that users can also just initialize the plotter and use the show method with the model and the scope directly. The mesh information is kept nicely in the model object rather than in the plotter, so you can use the plotter for other models too if needed, for example. 

I could implement something like this tha would behave as the previous API but I think it's kind of messy:

```python
def display(model, scope):
    PrimePlotter().show(model, scope)
```

What do you think?

RobPasMue replied:
I prefer your implementation to be honest - I am not a super fan of the plotter keeping relevant information that can get outdated so your implementation solves that issue. However, I believe it will also be up to @ansys/pyprimemesh-maintainers @ansys/pyprimemesh-admins to decide. =)

RobPasMue replied:
Or @ansys/pyprimemesh to be broad

RobPasMue commented on .github/workflows/ci_cd.yml:
```diff
@@ -21,6 +21,7 @@ env:
   PACKAGE_NAMESPACE: 'ansys.meshing.prime'
   DOCUMENTATION_CNAME: 'prime.docs.pyansys.com'
   RESET_IMAGE_CACHE: 0
+  PYANSYS_VISUALIZER_TESTMODE: true
```
Don't they have a conftest.py where we could define this behavior? I am tempted of changing it as well on other repos as well...

RobPasMue commented on src/ansys/meshing/prime/core/mesh.py:
```diff
@@ -0,0 +1,576 @@
+import numpy as np
+import pyvista as pv
+from ansys.tools.visualization_interface import MeshObjectPlot
+from beartype.typing import Dict, List, Union
```
Wait a second... are they using ``beartype``? If not, you have to switch to the ``typing`` module.

RobPasMue commented on src/ansys/meshing/prime/core/mesh.py:
```diff
@@ -0,0 +1,576 @@
+        Parameters
+        ----------
+        connectivity_results : Union[FaceConnectivityResults, EdgeConnectivityResults]
+            _description_
+        index : _type_
+            _description_
```
Descriptions are needed.

RobPasMue commented on src/ansys/meshing/prime/core/mesh.py:
```diff
@@ -0,0 +1,576 @@
+        """Get the polydata object of the faces.
+
+        Parameters
+        ----------
+        part_id : int
+            _description_
+        face_facet_res : FaceConnectivityResults
+            _description_
+        index : int
+            _description_
+
```
Same here... please review the codebase

RobPasMue commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,295 @@
+import numpy as np
+from ansys.tools.visualization_interface import Plotter
+from ansys.tools.visualization_interface.backends.pyvista import PyVistaBackend
+from beartype.typing import Any, Dict, List, Optional
```
Same issue here

RobPasMue commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,295 @@
+        """Add a Prime model to the plotter.
+
+        Parameters
+        ----------
+        model : _type_
+            _description_
+        scope : _type_, optional
+            _description_, by default None
+        """
```
Missing docstring definitions

RobPasMue commented on src/ansys/meshing/prime/graphics/widgets/images/show_ruler.png:
Is this one needed? I thought it comes by default with the viz interface

AlejandroFernandezLuces replied:
Nope, this one is custom for PyPrime. I think it can be added to the vizualization interface though, but do other libraries need this feature? I would prefer to avoid cluttering of the GUI as much as possible

RobPasMue reviewed (changes requested):
Some global changes requested from my side. Feel free to initiate a discussion on them :)

RobPasMue commented on src/ansys/meshing/prime/core/mesh.py:
```diff
@@ -488,7 +488,7 @@ def get_scoped_polydata(self, scope: "prime.ScopeDefinition"):
 
-    def update_pd(self, part_ids) -> Dict[int, Dict[str, List[(pv.PolyData, Part)]]]:
+    def update_pd(self, part_ids) -> Dict[int, Dict[str, list[(pv.PolyData, Part)]]]:
         """Update the polydata object of the mesh.
```
Watchout... I'm not sure what happens with Python 3.9 with ``list``.. Typing changed a lot from 3.9 to 3.10. You might have to import the old ``List`` style from typing

waltersma commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,300 @@
+from ansys.meshing.prime.graphics.widgets.color_by_type import ColorByTypeWidget
+from ansys.meshing.prime.graphics.widgets.hide_picked import HidePicked
+from ansys.meshing.prime.graphics.widgets.picked_info import PickedInfo
+from ansys.meshing.prime.graphics.widgets.toogle_edges import ToogleEdges
```
Hi @AlejandroFernandezLuces, multiple references to toogle, I believe this might be intended to be toggle? Thanks

RobPasMue reviewed (approved):
LGTM!

PipKat commented on examples/other/elbow_trame.py:
```diff
@@ -78,8 +78,9 @@
 # `use_trame` flag to True. When `display()` function executes, it will
 # pop a browser window with the visualizer open.
```
```suggestion
# ``use_trame`=True``. When the ``display()`` function executes, it
# opens the visualizer in a browser window.
```

PipKat commented on examples/other/pipe_tee.py:
```diff
@@ -85,8 +85,9 @@
 
 # Let us see what we read from the file. The file is an unmeshed
 # part as you would get after you imported from CAD.
```
```suggestion
# Read from the file, which is an unmeshed part
# like you would get after you imported from a CAD file.
```

PipKat commented on examples/other/pipe_tee.py:
```diff
@@ -158,7 +164,9 @@
 
 # Display mesh without unwanted EdgeZonelets. You can clearly see the
 # prism layers that were specified by the Prism control.
```
```suggestion
# Display mesh without unwanted edge zonelets. You can clearly see the
# prism layers that were specified by the Prism control.
```
In "Prism control," use double backticks around "Prism if it is a code entity.

PipKat commented on examples/other/pipe_tee.py:
```diff
@@ -158,7 +164,9 @@
+display.show()
 
 ###############################################################################
 # Write a cas file for use in the Fluent solver
```
```suggestion
# Write a CAS file for use in the Fluent solver
```

PipKat commented on src/ansys/meshing/prime/core/mesh.py:
```diff
@@ -0,0 +1,580 @@
+
+
+class DisplayMeshInfo:
+    """Contains the information of the mesh we want to display.
```
```suggestion
    """Contains the mesh information to display.
```

PipKat commented on src/ansys/meshing/prime/core/mesh.py:
```diff
@@ -0,0 +1,580 @@
+
+    Parameters
+    ----------
+    id : int
+        ID of the mesh.
+    part_id : int
+        ID of the part.
+    part_name : str
+        Name of the part.
+    zone_id : int
+        ID of the zone.
+    zone_name : str
+        Name of the zone.
+    display_mesh_type : DisplayMeshType
+        Type of mesh to display.
```
```suggestion
    id : int, default: 0
        ID of the mesh.
    part_id : int, default: 0
        ID of the part.
    part_name : str, default: None
        Name of the part.
    zone_id : int, default: 0
        ID of the zone.
    zone_name : str, default: None
        Name of the zone.
    display_mesh_type : DisplayMeshType, default: FACEZONELET
        Type of mesh to display.
```

PipKat commented on src/ansys/meshing/prime/core/mesh.py:
```diff
@@ -0,0 +1,580 @@
+        zone_name=None,
+        display_mesh_type=DisplayMeshType.FACEZONELET,
+    ) -> None:
+        """Initialize display mesh info."""
```
```suggestion
        """Initialize display mesh information."""
```

PipKat commented on src/ansys/meshing/prime/core/mesh.py:
```diff
@@ -0,0 +1,580 @@
+
+
+class Mesh(MeshInfo):
+    """Process the mesh for visualization in the GUI.
```
```suggestion
    """Processes the mesh for visualization in the GUI.
```
Class descriptions start with verbs ending in "s" or "es"

PipKat commented on src/ansys/meshing/prime/core/mesh.py:
```diff
@@ -0,0 +1,580 @@
+
+    @property
+    def model(self):
+        """Return the model."""
```
```suggestion
        """Model."""
```
Properties are described as nouns.

PipKat commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,353 @@
+
+    Parameters
+    ----------
+    use_trame : Optional[bool], optional.
+        Whether to use trame or not, by default None.
+    allow_picking : Optional[bool], optional.
+        Whether to allow picking or not, by default True.
```
```suggestion
    use_trame : Optional[bool], default: None
        Whether to use Trame.
    allow_picking : Optional[bool], default: True.
        Whether to allow picking.
```
I thought you were specifying the default value after the data type. Sorry if I'm remembering wrong! You can see that using this method often cuts down on words.

AlejandroFernandezLuces replied:
I think we are not consistent with this across PyAnsys libraries. I agree with you, we can use this method for this repo.

PipKat commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,353 @@
+
+    @property
+    def info_actor_map(self) -> Dict:
+        """Get the info actor map for picked info widget.
```
```suggestion
        """Get the info actor map for the selected information widget.
```
We should spell out words in descriptions.

PipKat commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,353 @@
+        Returns
+        -------
+        Dict
+            Info actor map.
```
```suggestion
            Information actor map.
```

PipKat commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,353 @@
+
+    @info_actor_map.setter
+    def info_actor_map(self, value: Dict) -> None:
+        """Set the info actor map for picked info widget.
```
```suggestion
        """Set the info actor map for the selected information widget.
```

PipKat commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,353 @@
+        Parameters
+        ----------
+        value : Dict
+            Info actor map.
```
```suggestion
            Information actor map.
```

PipKat commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,353 @@
+        Parameters
+        ----------
+        mesh_info : DisplayMeshInfo
+            Mesh info that generates an appropriate color.
```
```suggestion
            Mesh information that generates an appropriate color.
```

PipKat commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,353 @@
+        ----------
+        model : Model
+            Prime model to add.
+        scope : prime.ScopeDefinition, optional
+            Scope to show if any, by default None
```
```suggestion
        scope : prime.ScopeDefinition, default: None
            Scope to show, if any.
```

PipKat commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,353 @@
+        Parameters
+        ----------
+        plotting_list : List[Any]
+            List of objects you want to plot.
+        name_filter : str, default: None
+            Regular expression with the desired name or names you want to include in the plotter.
```
```suggestion
            List of objects to plot.
        name_filter : str, default: None
            Regular expression with the desired name or names to include in the plotter.
```

PipKat commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,353 @@
+
+        Parameters
+        ----------
+        plottable_object : Any, optional
+            Object to show, by default None.
+        screenshot : str, optional
+            Path to save a screenshot, by default None.
+        name_filter : bool, optional
+            Flag to filter the object, by default None.
+        plotting_options : dict
+            Additional plotting options the selected backend accepts.
```
```suggestion
        plottable_object : Any, default: None
            Object to show.
        screenshot : str, default: None
            Path to save a screenshot to.
        name_filter : bool, default: None
            Whether to filter the object.
        plotting_options : dict, default: None
            Additional plotting options that the selected backend accepts.
```
Is ``name_filter`` really a Boolean? It seems like it would be string?

AlejandroFernandezLuces replied:
This is a mistake, it is indeed an string.

PipKat commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,353 @@
+    ----------
+    model : prime.Model
+        Model to show.
+    use_trame : bool, optional
+        Whether to use the Trame visualizer. The default is ``False``.
```
```suggestion
    use_trame : bool, default: False
        Whether to use the Trame visualizer.```

PipKat commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,353 @@
+        self.model = model
+        self.use_trame = use_trame
+        warnings.warn(
+            "DeprecationWarning: Graphics module is deprecated. Use `PrimePlotter` instead."
```
```suggestion
            "DeprecationWarning: The `Graphics` class is deprecated. Use the `PrimePlotter` class instead."
```

PipKat commented on src/ansys/meshing/prime/graphics/plotter.py:
```diff
@@ -0,0 +1,353 @@
+
+        Parameters
+        ----------
+        parts : Any, optional
+            Parts to show. The default is ``None``.
+        update : bool, optional
+            Whether to update the display. The default is ``True``.
+        spline : bool, optional
+            Whether to use splines. The default is ``False``.
+        scope : prime.ScopeDefinition, optional
+            Scope of the parts. The default is ``None``.
```
```suggestion
        parts : Any, default: None
            Parts to show. 
        update : bool, default: True
            Whether to update the display.
        spline : bool, default: False
            Whether to use splines.
        scope : prime.ScopeDefinition, default: None
            Scope of the parts.
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/color_by_type.py:
```diff
@@ -0,0 +1,141 @@
+
+
+class ColorByTypeWidget(PlotterWidget):
+    """Initialize the color by type button widget.
```
```suggestion
    """Initializes the color by the type button widget.
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/color_by_type.py:
```diff
@@ -0,0 +1,141 @@
+class ColorByTypeWidget(PlotterWidget):
+    """Initialize the color by type button widget.
+
+    This widget allows the user to change the color of the mesh
+    based on the zone, zonelet or part.
```
```suggestion
    This widget lets you change the color of the mesh
    based on the zone, zonelet, or part.
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/color_by_type.py:
```diff
@@ -0,0 +1,141 @@
+
+        Parameters
+        ----------
+        color_type : ColorByType, optional
+            Color type to use, by default ColorByType.ZONE.
```
```suggestion
        color_type : ColorByType, default: ColorByType.ZONE
            Color type to use.
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/color_by_type.py:
```diff
@@ -0,0 +1,141 @@
+        color_type : ColorByType
+            Color type to use.
+        mesh_info : DisplayMeshInfo
+            Mesh info that generates an appropriate color.
```
```suggestion
            Mesh information that generates an appropriate color.
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/hide_picked.py:
```diff
@@ -0,0 +1,80 @@
+
+
+class HidePicked(PlotterWidget):
+    """Initialize the hide picked button widget.
```
```suggestion
    """Initializes the hide picked button widget.
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/hide_picked.py:
```diff
@@ -0,0 +1,80 @@
+    Parameters
+    ----------
+    prime_plotter : Plotter
+        Plotter where to apply this widget.
```
```suggestion
        Plotter to apply this widget to.
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/picked_info.py:
```diff
@@ -0,0 +1,120 @@
+
+
+class PickedInfo(PlotterWidget):
+    """Initialize the picked info button widget.
+
+    This widget allows the user to get information about the picked mesh objects.
```
```suggestion
    """Initializes the picked information button widget.

    This widget lets you get information about the picked mesh objects.
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/picked_info.py:
```diff
@@ -0,0 +1,120 @@
+    Parameters
+    ----------
+    prime_plotter : PrimePlotter
+        Plotter where to apply this widget.
```
```suggestion
        Plotter to apply this widget to.
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/picked_info.py:
```diff
@@ -0,0 +1,120 @@
+        self._info_actor_map = self.prime_plotter._info_actor_map
+
+    def info_message(self, mesh_info: DisplayMeshInfo) -> str:
+        """Return the information message for the selected mesh object.
```
```suggestion
        """Get the information message for the selected mesh object.
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/picked_info.py:
```diff
@@ -0,0 +1,120 @@
+        Parameters
+        ----------
+        mesh_info : DisplayMeshInfo
+            The mesh info object to print.
```
```suggestion
            Mesh information object to print.
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/picked_info.py:
```diff
@@ -0,0 +1,120 @@
+        return msg
+
+    def callback(self, state: bool) -> None:
+        """Define callback function for the button widget.
```
```suggestion
        """Define the callback function for the button widget.
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/picked_info.py:
```diff
@@ -0,0 +1,120 @@
+
+        Parameters
+        ----------
+        state : bool
```
```suggestion
        state : bool, default: None
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/toggle_edges.py:
```diff
@@ -0,0 +1,76 @@
+
+
+class ToggleEdges(PlotterWidget):
+    """Toggle the edges of the mesh objects.
```
```suggestion
    """Toggles the edges of the mesh objects.
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/toggle_edges.py:
```diff
@@ -0,0 +1,76 @@
+    Parameters
+    ----------
+    prime_plotter : PrimePlotter
+        Plotter where to apply this widget.
```
```suggestion
        Plotter to apply this widget to.
```

PipKat commented on src/ansys/meshing/prime/graphics/widgets/picked_info.py:
```diff
@@ -0,0 +1,120 @@
+        Parameters
+        ----------
+        state : bool
+            State of the button widget.
```
Not sure I or a user would understand what Boolean value to specify. Can the description be more comprehensive? Same for other occurrences of the ``state`` parameter.

AlejandroFernandezLuces replied:
Honestly, I documented this function, as well as all other `callback` methods so pre-commit wouldn't complain. These functions are not meant to be used by anyone directly. This function is called when the user presses a button in the GUI. So I don't think it's worth explaining this in detail. This `state` is really the internal state of the button, whether it is activated or not. 

If you think it still is worth to add more detail to this, let me know

PipKat replied:
@AlejandroFernandezLuces 
Based on your explanation, can we change the description to this:

Whether the button widget is activated.

That would make the format consistent with our other Boolean descriptions.

AlejandroFernandezLuces commented:
@waltersma I've fixed the issues I've found in the examples, and added interactive examples in the documentation. If you want to check out the rendered examples, you can download the documentation from the artifacts and run this:

```bash
python -m http.server -b localhost 8080
```

[video.webm](https://github.com/user-attachments/assets/d1966c81-5741-474b-8b7a-89a0a7000ac7)


If you are worried about the buttons appearing in the static scene, it will be fixed in the next release of the visualization tool.

waltersma commented:
> @waltersma I've fixed the issues I've found in the examples, and added interactive examples in the documentation. If you want to check out the rendered examples, you can download the documentation from the artifacts and run this:
> 
> ```shell
> python -m http.server -b localhost 8080
> ```
> 
>  video.webm 
> If you are worried about the buttons appearing in the static scene, it will be fixed in the next release of the visualization tool.

This looks good to me!

AlejandroFernandezLuces commented:
Thanks everyone for your reviews 😄
